cli/node/mod.rs

```diff
@@ -710,69 +712,117 @@ pub fn translate_cjs_to_esm(
+  }
+
+  let mut temp_var_count = 0;
+  let mut export_index = 0;
+  let mut handled_reexports: HashSet<String> = HashSet::default();
+
+  let mut source = vec![
+    r#"var window = undefined;"#.to_string(),
+    r#"const require = Deno[Deno.internal].require.Module.createRequire(import.meta.url);"#.to_string(),
+  ];
+
+  let analysis = perform_cjs_analysis(specifier.as_str(), media_type, code)?;
+
   let root_exports = analysis
     .exports
     .iter()
     .map(|s| s.as_str())
     .collect::<HashSet<_>>();
-  let mut temp_var_count = 0;
 
-  let mut source = vec![
-    r#"const require = Deno[Deno.internal].require.Module.createRequire(import.meta.url);"#.to_string(),
-  ];
+  // (request, referrer, is_top_level_reexport)
+  let mut reexports_to_handle = VecDeque::new();
+  for reexport in analysis.reexports {
+    reexports_to_handle.push_back((reexport, specifier.clone(), true));
+  }
+
+  while let Some((reexport, referrer, is_top_level_reexport)) =
+    reexports_to_handle.pop_front()
+  {
+    if handled_reexports.contains(&reexport) {
+      continue;
+    }
+
+    handled_reexports.insert(reexport.to_string());
 
-  // if there are reexports, handle them first
-  for (idx, reexport) in analysis.reexports.iter().enumerate() {
     // Firstly, resolve relate reexport specifier
     let resolved_reexport = resolve(
-      reexport,
-      specifier,
+      &reexport,
+      &referrer,
       // FIXME(bartlomieju): check if these conditions are okay, probably
       // should be `deno-require`, because `deno` is already used in `esm_resolver.rs`
       &["deno", "require", "default"],
       npm_resolver,
     )?;
     let reexport_specifier =
       ModuleSpecifier::from_file_path(&resolved_reexport).unwrap();
+    let referrer_filepath_str = referrer
+      .to_file_path()
+      .unwrap()
+      .to_string_lossy()
+      .to_string();
     // Secondly, read the source code from disk
     let reexport_file = file_fetcher.get_source(&reexport_specifier).unwrap();
-    // Now perform analysis again
+
     {
-      let parsed_source = deno_ast::parse_script(deno_ast::ParseParams {
-        specifier: reexport_specifier.to_string(),
-        text_info: deno_ast::SourceTextInfo::new(reexport_file.source),
-        media_type: reexport_file.media_type,
-        capture_tokens: true,
-        scope_analysis: false,
-        maybe_syntax: None,
-      })?;
-      let analysis = parsed_source.analyze_cjs();
-
-      source.push(format!(
-        "const reexport{} = require(\"{}\");",
-        idx, reexport
-      ));
-
-      for export in analysis.exports.iter().filter(|e| {
-        e.as_str() != "default" && !root_exports.contains(e.as_str())
-      }) {
+      let analysis = perform_cjs_analysis(
+        reexport_specifier.as_str(),
+        reexport_file.media_type,
+        reexport_file.source.to_string(),
+      )?;
+
+      for reexport in analysis.reexports {
+        reexports_to_handle.push_back((
```

Comment:
It seems this goes over the files like so:

```
1
  3
    5
      8
    6
  4
    7
2
```

I think it should be the following to match the execution order?

```
1
  2
    3
      4
    5
  6
    7
8
```

Reply:
Is there a specific reason for it? Checking with how it's implemented in Node it actually goes depth first

Reply:
I believe so it matches the code at runtime?

Reply:
I'll look into that

Reply:
Actually, I think the generated code handles all of this. Nevermind.